### PR TITLE
Utils Test for Cost Calculations

### PR DESCRIPTION
**Pull request by tolga-akan**
Crete a new general function in utilts test to check all the cost components in the problem 
TODO:
- [x] Make the function as general as possible (initially for the all assets for heat networks, later we can extend it to other commodifities)

Modify var-opex part for:
- [ ] pumppower cost calculation
- [ ]  influence of given electricity price profile

### CODE REVIEW

**Comment by tolga-akan** [comment]
@KobusVanRooyen Hi Kobus. I made a cost calculation test as general as possible. This function calculates the cost for assets that we use in heat networks.  Assets like electrolyzer, pump, ... are not included yet. I did not incorporate influence of  pumpower cost and electricity price profile yet. Shall I also include those? What do you think

File: tests/test_cold_demand.py [hunk 1]
```diff
@@ -127,23 +133,12 @@ def test_airco(self):
             input_timeseries_file="timeseries.csv",
         )
         results = heat_problem.extract_results()
-        parameters = heat_problem.parameters(0)
-        name_to_id_map = heat_problem.esdl_asset_name_to_id_map
-
-        hp_id = name_to_id_map["HeatPump_b97e"]
 
+        cost_calculation_test(heat_problem, results)
```
Comment (KobusVanRooyen) [inline tests/test_cold_demand.py]: 1. Will this test still pass if the _variable_operational_cost and parameters[f"{hp_id}.variable_operational_cost_coefficient"] might be zero for some reason? 
2. Based on the question above-> what about adding a hard check in the test case that one expects results[f"{hp_id}__variable_operational_cost"] to be > then .... (some sensible/large number and not 0).
3. Some hard coded check like this will be needed in each test case where it is expected that a value should exist even if cost_calculation_test(heat_problem, results) is used. What do you think?

File: tests/test_gas_electricity.py [hunk 2]
```diff
@@ -36,14 +35,13 @@ def test_gas_elect(self):
 
 
         Checks:
-        1. utils test: demand_matching_test, energy_conservation_test, heat_to_discharge_test,
-        electric_power_conservation_test, gas_pipes_head_loss_test
+        1. utils test: cost_calculation_test, demand_matching_test, energy_conservation_test,
```
Comment (KobusVanRooyen) [inline tests/test_gas_electricity.py]: Previously the TCO was checked, an not not anymore?

File: tests/test_elec_boiler.py [hunk 3]
```diff
@@ -95,20 +63,17 @@ def test_elec_heat_source_elec(self):
...
-        1. demand is matched
-        2. energy conservation in the network
-        3. heat to discharge
-        4. energy conservation over the heat and electricity commodity
-        5. ElectricBoiler cost components
+        1. cost calculation is checked
+        2. demand is matched
+        3. energy conservation in the network
+        4. heat to discharge
+        5. energy conservation over the heat and electricity commodity
```
Comment (KobusVanRooyen) [inline tests/test_elec_boiler.py]: why has item 5 been added? It is already mentioned in item 3?

File: tests/test_elec_boiler.py [hunk 4]
```diff
@@ -17,51 +18,17 @@
...
```
Comment (KobusVanRooyen) [inline tests/test_elec_boiler.py]: why has item 5 been added? It is already mentioned in item 3?

File: tests/test_elec_boiler.py [hunk 5]
```diff
@@ -80,6 +47,7 @@ def test_elec_heat_source_elec(self):
         parameters = heat_problem.parameters(0)
         name_to_id_map = heat_problem.esdl_asset_name_to_id_map
 
+        cost_calculation_test(heat_problem, results)
```
Comment (KobusVanRooyen) [inline tests/test_elec_boiler.py]: checks needed before the function call to make sure the specific cost items exists (> than sensible value)

File: tests/test_elec_boiler.py [hunk 6]
```diff
@@ -127,6 +92,7 @@ def test_heat_source_elec(self):
...
```
Comment (KobusVanRooyen) [inline tests/test_elec_boiler.py]: checks needed before the function call to make sure the specific cost items exists (> than sensible value)

File: tests/test_elec_boiler.py [hunk 7]
```diff
@@ -170,6 +134,7 @@ def test_air_water_hp_elec(self):
...
```
Comment (KobusVanRooyen) [inline tests/test_elec_boiler.py]: checks needed before the function call to make sure the specific cost items exists (> than sensible value)

File: tests/test_end_scenario_sizing.py [hunk 8]
```diff
@@ -111,6 +111,7 @@ def test_end_scenario_sizing(self):
         # Check that computation time is within expected bounds
 
         # Check whehter the heat demand is matched
+        cost_calculation_test(self.solution, self.results)
```
Comment (KobusVanRooyen) [inline tests/test_end_scenario_sizing.py]: checks needed before the function call to make sure the specific cost items exists (> than sensible value)

File: tests/test_gas_boiler.py [hunk 9]
```diff
@@ -82,6 +54,7 @@ def test_gas_heat_source_gas(self):
         gas_producer_id = name_to_id_map["GasProducer_82ec"]
         pipe_id = name_to_id_map["Pipe_a7b5"]
 
+        cost_calculation_test(heat_problem, results)
```
Comment (KobusVanRooyen) [inline tests/test_gas_boiler.py]: checks needed before the function call to make sure the specific cost items exists (> than sensible value)

File: tests/test_gas_boiler.py [hunk 10]
```diff
@@ -141,6 +111,7 @@ def test_heat_source_gas(self):
 
         gas_heater_id = name_to_id_map["GasHeater_f713"]
 
+        cost_calculation_test(solution, results)
```
Comment (KobusVanRooyen) [inline tests/test_gas_boiler.py]: checks needed before the function call to make sure the specific cost items exists (> than sensible value)

File: tests/test_gas_electricity.py [hunk 11]
```diff
@@ -76,12 +74,14 @@ def test_gas_elect(self):
         results_high_demand = solution_high_demand.extract_results()
 
         # Test: Utils_tests
+        cost_calculation_test(solution, results)
```
Comment (KobusVanRooyen) [inline tests/test_gas_electricity.py]: checks needed before the function call to make sure the specific cost items exists (> than sensible value)

File: tests/test_gas_electricity.py [hunk 12]
```diff
@@ -76,12 +74,14 @@ def test_gas_elect(self):
...
         demand_matching_test(solution, results)
         energy_conservation_test(solution, results)
         heat_to_discharge_test(solution, results)
         electric_power_conservation_test(solution, results)
         gas_pipes_head_loss_test(solution, results)
 
+        cost_calculation_test(solution_high_demand, results_high_demand)
```
Comment (KobusVanRooyen) [inline tests/test_gas_electricity.py]: checks needed before the function call to make sure the specific cost items exists (> than sensible value)

File: tests/test_gas_electricity.py [hunk 13]
```diff
@@ -138,93 +138,6 @@ def test_gas_elect(self):
...
-                for ii in range(1, len(solution.times())):
-                    variable_operational_cost += (
-                        var_op_costs * nominator_vector[ii] * timesteps_hr[ii - 1] / factor
-                    )
-            np.testing.assert_allclose(
-                variable_operational_cost, results[f"{asset_id}__variable_operational_cost"]
-            )
-            total_opex += variable_operational_cost
-
-        np.testing.assert_allclose(
```
Comment (KobusVanRooyen) [inline tests/test_gas_electricity.py:224]: The obj is not tested anymore. Please add it to the new function "cost_calculation_test" in the test utils. But this check of obj value is not always required. So create a new input for the function "cost_calculation_test", something like check_objective_function and make the default value False. 

So for this specific test case the input variable will be set to check_objective_function=True. Then for all the other test case check_objective_function does not have to be set, if the objective function value is not tested

File: tests/utils_tests.py [hunk 14]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
                 np.testing.assert_allclose(
                     results[f"{pipe}.dH"][itime], -dh_manual_linear, atol=atol
                 )
+
+
+def _get_esdl_scaled_cost(cost_esdl_info, uni_enum):
```
Comment (KobusVanRooyen) [inline tests/utils_tests.py:605]: please add docstring for this function

File: tests/utils_tests.py [hunk 15]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+            per_scaler = 1.0e6
+        else:
+            raise RuntimeWarning(f"{per_multiplier} is not supported as unit per multiplier.")
+    else:  # CUBIC_METRE or METRE or NONE
+        per_scaler = 1.0
+    cost_scaled = costs_esdl / per_scaler
+    return cost_scaled
+
+
+def cost_calculation_test(solution, results, atol=1e-8):
```
Comment (KobusVanRooyen) [inline tests/utils_tests.py:645]: please add a docstring for this function

File: tests/utils_tests.py [hunk 16]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+    for asset in assets:
+        esdl_asset = solution.esdl_assets[asset]
+        costs_esdl_asset = esdl_asset.attributes["costInformation"]
+        if costs_esdl_asset is None:
+            continue
+
+        # Investment Cost
+        investment_cost = 0.0
+        if asset not in solution.energy_system_components.get("gas_pipe", []):
+            investment_cost_info = _get_esdl_scaled_cost(
```
Comment (KobusVanRooyen) [inline tests/utils_tests.py:712]: Can we add testing for the the cost coefficient itself,  to check that the value in the esdl (which investment_cost_info) is the same as the value stored in the MESIDO parameters (whic his parameters[f"{asset}.variable_operational_cost_coefficient"] in this case)

The benefit is that we are also checking if we have parsed and save the actual cost coef correctly. Can you add this to all the other cost coefficients in this function please

File: tests/utils_tests.py [hunk 17]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+            elif asset in solution.energy_system_components.get("airco", []):
+                nominator_vector = results[f"{asset}.Heat_airco"]
+                for ii in range(1, len(solution.times())):
+                    variable_operational_cost += (
+                        var_op_costs_esdl * nominator_vector[ii] * timesteps_hr[ii - 1]
+                    )
+        np.testing.assert_allclose(
+            variable_operational_cost, results[f"{asset}__variable_operational_cost"]
+        )
+        total_opex += variable_operational_cost
```
Comment (KobusVanRooyen) [inline tests/utils_tests.py]: please add a check for the obj function, but only checking it if check_objective_function=True

File: tests/test_cold_demand.py (hunk 1 above)
Comment (FJanssen-TNO) [inline tests/test_cold_demand.py]: The first tests should be related to the physics, heat_demand_matching energy_conservation and heat_to_discharge tests as these tests also tell us whether the majority of the code is working. If there already an issue occurs then the costs calucation might also fail, but that is then not the cause. 
This is for all the tests

File: tests/utils_tests.py [hunk 18]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+                else:
+                    nominator_vector = heat_source
+
+                for ii in range(1, len(solution.times())):
+                    variable_operational_cost += (
+                        var_op_costs_esdl
+                        * nominator_vector[ii]
+                        * timesteps_hr[ii - 1]
+                        / denominator
+                    )
```
Comment (FJanssen-TNO) [inline tests/utils_tests.py]: No for loop for the times is need, like discussed before. also in the elif below.

File: tests/utils_tests.py [hunk 19]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+        *solution.energy_system_components.get("heat_exchanger", []),
+        *solution.energy_system_components.get("heat_buffer", []),
+        *solution.energy_system_components.get("gas_source", []),
+        *solution.energy_system_components.get("electricity_source", []),
+        *solution.energy_system_components.get("airco", []),
+        *demand_assets,
+        *transport_assets,
+    ]
+
+    pipe_classes = [
```
Comment (FJanssen-TNO) [inline tests/utils_tests.py]: rename to gas_pipe_classes

File: tests/utils_tests.py [hunk 20]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+            if parameters[f"{asset}.diameter"] > 0:
+                for iter in range(len(pipe_classes)):
+                    if pipe_classes[iter].inner_diameter == parameters[f"{asset}.diameter"]:
+                        investment_cost = (
+                            pipe_classes[iter].investment_costs * parameters[f"{asset}.length"]
+                        )
+        elif asset in solution.energy_system_components.get("heat_pipe", []):
+            investment_cost = investment_cost_info * parameters[f"{asset}.length"]
+        elif asset in solution.energy_system_components.get("electricity_cable", []):
+            investment_cost = investment_cost_info * parameters[f"{asset}.length"]
```
Comment (FJanssen-TNO) [inline tests/utils_tests.py]: combine these elif statements

File: tests/utils_tests.py [hunk 21]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+            investment_cost = investment_cost_info * heat_buffer_volume
+        else:
+            investment_cost = investment_cost_info * results[f"{asset}__max_size"]
+        total_capex += investment_cost
+        np.testing.assert_allclose(
+            investment_cost, results[f"{asset}__investment_cost"], atol=1.0e-6
+        )
+
+        # Installation Cost
+        if asset not in transport_assets:
```
Comment (FJanssen-TNO) [inline tests/utils_tests.py]: Why are transport_assets excluded? Wouldn't it be better to check that it is 0.0 or None and thus also 0.0

File: tests/utils_tests.py [hunk 22]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+                *solution.energy_system_components.get("low_temperature_ates", []),
+            ]:
+                nominator_vector = (
+                    results[f"{asset}.Heat_flow_charging"]
+                    + results[f"{asset}.Heat_flow_discharging"]
+                )
+                for ii in range(1, len(solution.times())):
+                    variable_operational_cost += (
+                        var_op_costs_esdl * nominator_vector[ii] * timesteps_hr[ii - 1]
+                    )
```
Comment (FJanssen-TNO) [inline tests/utils_tests.py]: no for loop needed

File: tests/utils_tests.py [hunk 23]
```diff
@@ -596,3 +600,193 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+                    nominator_vector = heat_source
+
+                for ii in range(1, len(solution.times())):
+                    variable_operational_cost += (
+                        var_op_costs_esdl
+                        * nominator_vector[ii]
+                        * timesteps_hr[ii - 1]
+                        / denominator
+                    )
+            elif asset in solution.energy_system_components.get("airco", []):
```
Comment (FJanssen-TNO) [inline tests/utils_tests.py:841]: Missing an else statement. Now there are several elif statemetns, but those do not cover all the non transport assets, so without realising you are missing out on several assets. I believe all those assets should be calculated and checked. Also the heat demand etc.

File: tests/utils_tests.py (hunk 22 above)
Comment (FJanssen-TNO) [inline tests/utils_tests.py]: Can be moved down and generalized instead of being repeated several times, all variable opex is calculated with var_op_costs_esdl * nominator*dt/denominator.

File: tests/utils_tests.py (hunk 17 above)
Comment (FJanssen-TNO) [reply tests/utils_tests.py]: For a check on the objective function, you need to check the state of the asset (whether it is enabled or optional)

File: tests/test_end_scenario_sizing.py
Comment (FJanssen-TNO) [inline tests/test_end_scenario_sizing.py:1]: I think the costs check on demands that is covered in this test can also be removed once it is included in cost_calculation_test

File: tests/test_cold_demand.py
Comment (FJanssen-TNO) [inline tests/test_cold_demand.py:1]: test_max_size_and_optional_assets.py also contains many costs checks, some hard-coded values, those you should keep, but the once where it checks the calculation you should remove and add the cost_calculation_test

File: tests/test_cold_demand.py (hunk 1 above)
Comment (tolga-akan) [reply tests/test_cold_demand.py]: Yes. This is a good point. I will add additional check to all tests (if needed) to ensure regarding cost component is non zero (larger than a large enough positive value)

File: tests/test_elec_boiler.py (hunk 4 above)
Comment (tolga-akan) [reply tests/test_elec_boiler.py]: it should be "electric_power_conservation_test". now updated

File: tests/utils_tests.py (hunk 23 above)
Comment (KobusVanRooyen) [reply tests/utils_tests.py:841]: I would also prefer an "else .... error exit" to cater for cases when new assets are added, which are not included in the test. This way we are made aware of an issue, else we might think the test is doing something where is might not be doing/testing anything

**Comment by KobusVanRooyen** [comment]
@tolga-akan review completed

File: tests/test_elec_boiler.py (hunk 3 above)
Comment (tolga-akan) [reply tests/test_elec_boiler.py]: actually that line was refering to  the followign check i the test:         np.testing.assert_array_less(0.0, results[f"{e_boiler_id}.Heat_source"])
        np.testing.assert_allclose(
            parameters[f"{e_boiler_id}.efficiency"] * results[f"{e_boiler_id}.Power_consumed"],
            results[f"{e_boiler_id}.Heat_source"],
        )

However, you are right that  item 3 is already hinting the same. Now I removed item 5

File: tests/test_elec_boiler.py (hunk 5 above)
Comment (tolga-akan) [reply tests/test_elec_boiler.py]: Check that ensures cost components of elec heat sourc elec are non-zero is added

File: tests/test_elec_boiler.py (hunk 6 above)
Comment (tolga-akan) [reply tests/test_elec_boiler.py]: Added -> Check cost components of heat sourc elec are non-zero

File: tests/test_elec_boiler.py (hunk 7 above)
Comment (tolga-akan) [reply tests/test_elec_boiler.py]: required check is added

File: tests/test_end_scenario_sizing.py (hunk 8 above)
Comment (tolga-akan) [reply tests/test_end_scenario_sizing.py]: required checks are added

File: tests/test_gas_boiler.py (hunk 9 above)
Comment (tolga-akan) [reply tests/test_gas_boiler.py]: required checks are added

File: tests/test_gas_boiler.py (hunk 10 above)
Comment (tolga-akan) [reply tests/test_gas_boiler.py]: required checks are added

File: tests/test_gas_electricity.py (hunk 11 above)
Comment (tolga-akan) [reply tests/test_gas_electricity.py]: required checks area placed

File: tests/test_gas_electricity.py (hunk 12 above)
Comment (tolga-akan) [reply tests/test_gas_electricity.py]: required checks area placed

File: tests/utils_tests.py (hunk 14 above)
Comment (tolga-akan) [reply tests/utils_tests.py:605]: docstring is added

File: tests/utils_tests.py (hunk 15 above)
Comment (tolga-akan) [reply tests/utils_tests.py:645]: docstring is added

File: tests/utils_tests.py (hunk 19 above)
Comment (tolga-akan) [reply tests/utils_tests.py]: renamed to gas_pipe_classes

File: tests/utils_tests.py (hunk 16 above)
Comment (tolga-akan) [reply tests/utils_tests.py:712]: I added the suggested checks except "installation cost"and investemt cost of heat pipes. Because, there is no variable called  installation_cost_coefficient. I could not figured out the reason but I see that parameters[f"{asset}.investment_cost_coefficient"] and investment_cost_info are different from each other.

File: tests/utils_tests.py (hunk 20 above)
Comment (tolga-akan) [reply tests/utils_tests.py]: combined

File: tests/utils_tests.py (hunk 21 above)
Comment (tolga-akan) [reply tests/utils_tests.py]: Yes, that would be better. I will add check for also transport assets

File: tests/utils_tests.py (hunk 22 above)
Comment (tolga-akan) [reply tests/utils_tests.py]: for loop is replaced by multiplication of vectors

Comment (tolga-akan) [reply tests/utils_tests.py]: I agree. now it is moved down

File: tests/utils_tests.py (hunk 18 above)
Comment (tolga-akan) [reply tests/utils_tests.py]: for loops are removed and generalized by vector multiplication at the bottom

File: tests/test_cold_demand.py (hunk 1 above)
Comment (tolga-akan) [reply tests/test_cold_demand.py]: I agree. order of the cost test is changed.

**Review by tolga-akan — COMMENTED** [review]
@KobusVanRooyen and @FJanssen-TNO I tried to address the comments as much as possible but it is not complete yet. There are still missing points that I could not respond due to lack of time.

File: tests/utils_tests.py [hunk 24]
```diff
@@ -596,3 +600,257 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+                    nominator_vector = results[f"{asset}.Power_consumed"]  # [W]
+                else:
+                    nominator_vector = heat_source
+            elif asset in solution.energy_system_components.get("airco", []):
+                nominator_vector = results[f"{asset}.Heat_airco"]
+
+            variable_operational_cost = sum(
+                var_op_costs_esdl
+                * nominator_vector[1 : len(solution.times())]
+                * timesteps_hr[: len(solution.times()) - 1]
```
Comment (KobusVanRooyen) [inline tests/utils_tests.py]: for timesteps_hr[: len(solution.times()) - 1] you do need to use the index. Use timesteps_hr without specifying the index

File: tests/utils_tests.py [hunk 25]
```diff
@@ -596,3 +600,257 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+                ]:
+                    nominator_vector = results[f"{asset}.Power_consumed"]  # [W]
+                else:
+                    nominator_vector = heat_source
+            elif asset in solution.energy_system_components.get("airco", []):
+                nominator_vector = results[f"{asset}.Heat_airco"]
+
+            variable_operational_cost = sum(
+                var_op_costs_esdl
+                * nominator_vector[1 : len(solution.times())]
```
Comment (KobusVanRooyen) [inline tests/utils_tests.py]: nominator_vector[1 : len(solution.times())] can also be replaced by nominator_vector[1: ] ?

File: tests/utils_tests.py (hunk 16 above)
Comment (KobusVanRooyen) [reply tests/utils_tests.py:782]: When you saw this difference for heat_pipe was it when the pipe was optimized? Because then the cost in the esdl might not be the same as the cost in the result becasue we use the updated cost (json file-> pipe catalog) and not the cost in the esdl

File: tests/utils_tests.py (hunk 24 above)
Comment (tolga-akan) [reply tests/utils_tests.py]: You are right. Indexing is not needed. Now it is removed.

File: tests/utils_tests.py (hunk 25 above)
Comment (tolga-akan) [reply tests/utils_tests.py]: Yes you are right. the length of solution.times() and nominator_vector are already equal to each other. Suggestion is implemented now.

File: tests/test_gas_electricity.py (hunk 2 above)
Comment (tolga-akan) [reply tests/test_gas_electricity.py]: now cost calculation functions is modified. Hence, objective_value of the problem is verified with the cost

File: tests/test_gas_electricity.py (hunk 13 above)
Comment (tolga-akan) [reply tests/test_gas_electricity.py:224]: Your suggestion is implemented now.

File: tests/utils_tests.py (hunk 17 above)
Comment (tolga-akan) [reply tests/utils_tests.py]: Suggestion is implemented. Now each assets is checked to decide to calculate the contributrion of assets to the objective value

File: tests/utils_tests.py (hunk 23 above)
Comment (tolga-akan) [reply tests/utils_tests.py:917]: else is added with assertionerror

File: tests/utils_tests.py (hunk 16 above)
Comment (tolga-akan) [reply tests/utils_tests.py:782]: Yes, correct. This only occurs for optional pipe. Now _edr_pipes.json is used for optional pipes

File: tests/test_end_scenario_sizing.py
Comment (tolga-akan) [reply tests/test_end_scenario_sizing.py:1]: currently cost_calculation_test function in utils_tests is already taking the investment cost of demand assets into account. Hence, I have removed explicit calculations from the test since all are already covered in cost_calculation_Test

File: tests/utils_tests.py (hunk 23 above)
Comment (tolga-akan) [reply tests/utils_tests.py:917]: Heat demand cost calculation is added

File: tests/test_cold_demand.py
Comment (tolga-akan) [reply tests/test_cold_demand.py:1]: test_max_size_and_optional_assets.py solves 2 heat problem. I have placed cost_calculation_test function to only the 2nd one. Because, I have observe a wierd result when I check the investment cost of Pipe_9768_ret'. For this pipe results[f"{asset}__investment_cost"] returns 313257.47 this is possible only possible if the pipe DN is 200. Whereas the cost calculation function calculates the cost of this pipe for verification (it calls the pipe diameter from solution.esdl_assets[asset] which is DN450) as 656578.59. This is the cost for DN450. I think the reason is that test_max_size_and_optional_assets.py calls HeatProblem to formulate the problem. And this problem is making this pipe optional.

**Comment by tolga-akan** [comment]
@KobusVanRooyen @FJanssen-TNO Ready to be reviewed

File: CHANGELOG.md [hunk 26]
```diff
@@ -8,6 +8,7 @@
 - Generic methods to create equality and inequality constraints using big-M formulation.
 - First implementation of a temperature profile for a producer. (experimental)
 - Test is added to check airco sizing and cost calculation
+- cost_calculation_test is added into utils_Tests
```
Comment (KobusVanRooyen) [inline CHANGELOG.md]: As discussed yesterday. changes in test cases / tests do not have to go in the changelog

File: tests/test_gas_electricity.py [hunk 27]
```diff
@@ -138,93 +154,6 @@ def test_gas_elect(self):
...
-                investment_cost = investment_cost_info * parameters[f"{asset_id}.length"]
-            elif asset_id in solution.energy_system_components["gas_pipe"]:
-                if parameters[f"{asset_id}.diameter"] > 0:
-                    for iter in range(len(pipe_classes)):
-                        if pipe_classes[iter].inner_diameter == parameters[f"{asset_id}.diameter"]:
-                            investment_cost = (
-                                pipe_classes[iter].investment_costs
-                                * parameters[f"{asset_id}.length"]
-                            )
-            total_capex += investment_cost
```
Comment (KobusVanRooyen) [inline tests/test_gas_electricity.py:177]: I do not see any hard coded checks to make sure the investment cost exists > ...

File: tests/test_gas_electricity.py [hunk 28]
```diff
@@ -138,93 +154,6 @@ def test_gas_elect(self):
...
-            np.testing.assert_allclose(
-                investment_cost, results[f"{asset_id}__investment_cost"], atol=1.0e-6
-            )
-
-            # installation cost
-            if asset_id in solution.energy_system_components["heat_source"]:
```
Comment (KobusVanRooyen) [inline tests/test_gas_electricity.py:183]: I do not see any hard coded checks to make sure the installation cost exists > ...

Please do a full check that all values are hard coded check in all the places that the cost utils it used before asking for a PR review again. Else me/Femke have to check everything in detail

File: tests/utils_tests.py [hunk 29]
```diff
@@ -635,3 +641,290 @@ def gas_pipes_head_loss_test(solution, results, atol=1e-12):
...
+                        * parameters[f"{asset}.rho"]
+                        * parameters[f"{asset}.dT"]
+                    ),
+                    investment_cost_info,
+                )
+            elif (
+                asset in solution.energy_system_components.get("heat_pipe", [])
+                and esdl_asset.attributes["state"] == esdl.AssetStateEnum.OPTIONAL
+            ):
+                investment_cost_info_opt_pipe = edr_pipes[
```
Comment (KobusVanRooyen) [inline tests/utils_tests.py]: I do not think this is working for an optional pipe, as discussed over teams

File: tests/test_end_scenario_sizing.py [hunk 30]
```diff
@@ -174,17 +173,6 @@ def test_end_scenario_sizing(self):
                 heat_buffer, heat_flow_charging - heat_flow_discharging, atol=1.0
             )
 
-        obj = self.calculate_objective_value_end_scenario_sizing_all_optional(self.solution)
-        excluded_costs_in_obj = self.calculate_heat_demand_costs_end_scenario_sizing()  # Fixed
```
Comment (KobusVanRooyen) [inline tests/test_end_scenario_sizing.py:178]: where is the hard coded check now to ensure these values exist? It is for a heating demand correct? Installation or investment cost?

File: tests/test_end_scenario_sizing.py
Comment (KobusVanRooyen) [reply tests/test_end_scenario_sizing.py:1]: Why are you mentioning investment cost? Is it not installation cost that should be checked for the heat demand? Is the installation cost checked for heating demand in utils?

Please check these things yourself. Installation vs investment costs.... before asking for a PR review.

File: tests/test_cold_demand.py
Comment (KobusVanRooyen) [reply tests/test_cold_demand.py:1]: Please fix your utils test. It should be working for optional pipes as well

**Comment by KobusVanRooyen** [comment]
@tolga-akan review has been completed

File: tests/test_cold_demand.py
Comment (tolga-akan) [reply tests/test_cold_demand.py:1]: Update: Npw, above mentioned problem is addressed. Previously I assumed that I can get access to the optimized DN cost via solution.esdl_assets[asset]. However, after discussing with Kobus I have realized that solution.esdl_assets[asset] returns the innitial DN size of optional pipe before the optimization. Now, I get the cost of the optimized pipe via different method. If the pipe is optional, I first get the resulting diameter of the pipe from parameters[f"{asset}.diameter"] and later I search the corresponding cost of this diameter from edr_pipes dictionary. This resolved the problem. Hence, I can freely use cost_calculation_test function in test_max_size_and_optional_assets.py.

File: CHANGELOG.md (hunk 26 above)
Comment (tolga-akan) [reply CHANGELOG.md]: line in the changelog is removed

File: tests/test_gas_electricity.py (hunk 27 above)
Comment (tolga-akan) [reply tests/test_gas_electricity.py:177]: all the required hard coded checks are added now

File: tests/test_end_scenario_sizing.py
Comment (tolga-akan) [reply tests/test_end_scenario_sizing.py:1]: I could be more precise at responding to the previous comment. cost_calculation_test function in utils_tests is already taking the investment cost **and installation cost** of demand assets into account. 

now I added an additional manual check to verify installation cost of heat demands are non-zero before calling cost_calculation_test  function.

File: tests/test_gas_electricity.py (hunk 28 above)
Comment (tolga-akan) [reply tests/test_gas_electricity.py:183]: hard coded checks are added

File: tests/utils_tests.py (hunk 29 above)
Comment (tolga-akan) [reply tests/utils_tests.py]: Now I made required update. I created a new function. If the heat pipe is optional, the new function is used.  This function searches the EDR pipes dictionary for a Steel-S1 pipe with the specified inner diameter and returns its investment cost.

File: tests/test_end_scenario_sizing.py (hunk 30 above)
Comment (tolga-akan) [reply tests/test_end_scenario_sizing.py:178]: Please correct me if I have misunderstood your questions. Are you asking me to add a hard‑coded check to ensure that the heat demand installation cost is greater than zero? If so, I have now added such a check to verify that the heat demand installation cost is strictly positive.

**Review by KobusVanRooyen — APPROVED** [review]

**Comment by KobusVanRooyen** [comment]
@tolga-akan approved
